webapp/channels/src/utils/desktop_api.ts

```diff
@@ -0,0 +1,288 @@
+        if (window.desktopAPI?.getAppInfo) {
+            return window.desktopAPI.getAppInfo();
+        }
+
+        return this.invokeWithMessaging<void, {name: string; version: string}>(
+            'webapp-ready',
+            undefined,
+            'register-desktop',
+        );
+    };
+
+    /**
+     * @deprecated
+     */
+    private postMessageListener = ({origin, data: {type, message}}: {origin: string; data: {type: string; message: unknown}}) => {
+        if (origin !== window.location.origin) {
+            return;
+        }
+
+        const listeners = this.postMessageListeners?.get(type);
+        if (!listeners) {
+            return;
+        }
+
+        listeners.forEach((listener) => {
+            listener(message);
+        });
+    };
+
+    /**
+     * @deprecated
+     */
+    addPostMessageListener = (channel: string, listener: (message: any) => void) => {
+        if (this.postMessageListeners?.has(channel)) {
+            this.postMessageListeners.set(channel, this.postMessageListeners.get(channel)!.add(listener));
+        } else {
+            this.postMessageListeners?.set(channel, new Set([listener]));
+        }
+    };
+
+    /**
+     * @deprecated
+     */
+    removePostMessageListener = (channel: string, listener: (message: any) => void) => {
+        const set = this.postMessageListeners?.get(channel);
+        set?.delete(listener);
+        if (set?.size) {
+            this.postMessageListeners?.set(channel, set);
+        } else {
+            this.postMessageListeners?.delete(channel);
+        }
+    };
+
+    /**
+     * @deprecated
+     */
+    invokeWithMessaging = <T, T2>(
+        sendChannel: string,
+        sendData?: T,
+        receiveChannel?: string,
+    ) => {
+        return new Promise<T2>((resolve) => {
+            /* create and register a temporary listener if necessary */
+            const response = ({origin, data: {type, message}}: {origin: string; data: {type: string; message: T2}}) => {
+                /* ignore messages from other frames */
+                if (origin !== window.location.origin) {
+                    return;
+                }
+
+                /* ignore messages from other channels */
+                if (type !== (receiveChannel ?? sendChannel)) {
+                    return;
+                }
+
+                /* clean up listener and resolve */
+                window.removeEventListener('message', response);
+                resolve(message);
+            };
+
+            window.addEventListener('message', response);
+            window.postMessage(
+                {type: sendChannel, message: sendData},
+                window.location.origin,
+            );
+        });
+    };
+}
+
+const desktopApp = new DesktopApp();
+export default desktopApp;
+
+/*******************************
+ * API Methods
+ *******************************/
+
+/**
+ * Invokes
+ */
+
+export const getBrowserHistoryStatus = async () => {
```

Comment:
In a similar vein, I wonder if making `desktopAPI` a field of the `DesktopApp` class would be a good idea so that we could inject a mock easier without having to change things on the window. That's probably a moot point though as long as `DesktopApp` is a singleton

Reply:
Also went back and forth on this. I notice we don't really use classes much in the web app so I didn't want to force most of the API methods to be a member (singleton or not)

But I felt I needed one to store a few fields (without just exporting variables to be read), so I'm kinda stuck in the middle as to how to proceed.

My gut reaction here is to actually split up the `DesktopApp` class and the API methods into 2 seperate files, thoughts?

Reply:
I'd actually prefer the API methods to be instance methods of a class. I think the desktop API is similar enough to something like `Client4` or `WebSocketClient` that keeping them in a class makes sense to me conceptually. Plus, if we end up wanting to add internal state to the API code or doing something fancy, starting with a class encapsulates that and lets us make any changes without affecting callers.

As for if that should be in `DestkopApp` or in a separate class, I'm 0/5 on that. Keeping them together does mean exposing more areas of the code to the initialization and any lower level logic (like querying the desktop version directly), but it might make sense anyway. I feel similarly about if they should be in two different files or not, but that's only possible if they're not in a single class.

Reply:
If we're feeling that it should be a class member then I'm good with that, definitely feels cleaner to me. That way we'll just work with this one object. I'll push everything into the class.